vaadin-virtual-list-flow-parent/vaadin-virtual-list-flow/src/main/java/com/vaadin/flow/component/virtuallist/VirtualList.java

```diff
@@ -209,16 +201,33 @@ public void setRenderer(ValueProvider<T, String> valueProvider) {
     public void setRenderer(Renderer<T> renderer) {
         Objects.requireNonNull(renderer, "The renderer must not be null");
 
-        if (dataGeneratorRegistration != null) {
-            dataGeneratorRegistration.remove();
-            dataGeneratorRegistration = null;
+        rendererRegistrations.forEach(Registration::remove);
+        rendererRegistrations.clear();
+
+        Rendering<T> rendering;
+        if (renderer instanceof LitRenderer) {
+            // LitRenderer
+            if (template.getParent() != null) {
+                getElement().removeChild(template);
+            }
+            rendering = renderer.render(getElement(),
+                    dataCommunicator.getKeyMapper());
+        } else {
+            // TemplateRenderer or ComponentRenderer
+            if (template.getParent() == null) {
+                getElement().appendChild(template);
+            }
+            rendering = renderer.render(getElement(),
+                    dataCommunicator.getKeyMapper(), template);
```

Comment:
A `<template>` and a `renderer` function (one not assigned by `vaadin-template-renderer`) can't exist on a component at the same time.

This logic removes the `<template>` element from the host in case a `LitRenderer`'s `renderer` function is to be assigned. And respectively, restores the `<template>` element when a `TemplateRenderer` or a `ComponentRenderer` is set.